fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java

```diff
@@ -360,39 +362,86 @@ public PlanFragment visitAbstractPhysicalSort(AbstractPhysicalSort<Plan> sort, P
         PlanFragment rightFragment = hashJoin.child(1).accept(this, context);
         PlanFragment leftFragment = hashJoin.child(0).accept(this, context);
         PlanNode leftFragmentPlanRoot = leftFragment.getPlanRoot();
         PlanNode rightFragmentPlanRoot = rightFragment.getPlanRoot();
         JoinType joinType = hashJoin.getJoinType();
 
-        if (JoinUtils.shouldNestedLoopJoin(hashJoin)) {
-            throw new RuntimeException("Physical hash join could not execute without equal join condition.");
-        } else {
-            Expression eqJoinExpression = hashJoin.getCondition().get();
-            List<Expr> execEqConjunctList = ExpressionUtils.extractConjunction(eqJoinExpression).stream()
-                    .map(EqualTo.class::cast)
-                    .map(e -> swapEqualToForChildrenOrder(e, hashJoin.left().getOutput()))
-                    .map(e -> ExpressionTranslator.translate(e, context))
-                    .collect(Collectors.toList());
+        Expression eqJoinExpression = hashJoin.getCondition().get();
+        List<Expr> execEqConjunctList = ExpressionUtils.extractConjunction(eqJoinExpression).stream()
+                .map(EqualTo.class::cast)
+                .map(e -> swapEqualToForChildrenOrder(e, hashJoin.left().getOutput()))
+                .map(e -> ExpressionTranslator.translate(e, context))
+                .collect(Collectors.toList());
 
-            TupleDescriptor outputDescriptor = context.generateTupleDesc();
-            List<Expr> srcToOutput = hashJoin.getOutput().stream()
-                    .map(SlotReference.class::cast)
-                    .peek(s -> context.createSlotDesc(outputDescriptor, s))
-                    .map(e -> ExpressionTranslator.translate(e, context))
-                    .collect(Collectors.toList());
+        TupleDescriptor outputDescriptor = context.generateTupleDesc();
+        List<Expr> srcToOutput = hashJoin.getOutput().stream()
+                .map(SlotReference.class::cast)
+                .peek(s -> context.createSlotDesc(outputDescriptor, s))
+                .map(e -> ExpressionTranslator.translate(e, context))
+                .collect(Collectors.toList());
 
-            HashJoinNode hashJoinNode = new HashJoinNode(context.nextPlanNodeId(), leftFragmentPlanRoot,
-                    rightFragmentPlanRoot, JoinType.toJoinOperator(joinType), execEqConjunctList, Lists.newArrayList(),
-                    srcToOutput, outputDescriptor, outputDescriptor);
+        HashJoinNode hashJoinNode = new HashJoinNode(context.nextPlanNodeId(), leftFragmentPlanRoot,
+                rightFragmentPlanRoot, JoinType.toJoinOperator(joinType), execEqConjunctList, Lists.newArrayList(),
+                srcToOutput, outputDescriptor, outputDescriptor);
 
+        // TODO: use distribution info from physical properties.
+        boolean doBroadcast = !JoinUtils.onlyShuffle(hashJoin);
```

Comment:
This is a little bit aggressive. Shuffle Join is much safer than Broadcast Join. Broadcast join cause many BE OOM.
And hence, it would be better to choose shuffle join, if both join type are candidates, especially on TPC-H benchmark.

Reply:
This is draft version and waiting for #11666. Specific join selection strategy would use enforce framework then.